### DISCUSSION TITLE
Questions about processing WhatsApp backups found into Apple or Google Cloud Warrant Returns

### DISCUSSION THREAD

**Discussion by dafeol** [sepinf-inc/IPED#1657] [Q&A] [answered]
Boa tarde!
Estou com dados de duas nuvem, uma Google e outra da Apple. Minha dúvida é a respeito da leitura feita pelo IPED dos backups do Whatsapp. É necessário alguma configuração específica antes de realizar a indexação dos arquivos baixados destas nuvens? Ou o IPED por padrão já consegue ler estes backups e montar os diálogos da mesma forma que o CELLEBRITE?

Abaixo tenho por exemplos os arquivos que foram gerados após descriptografar e descompactar a nuvem do usuário, porém após a indexação, o IPED não me trás os diálogos:
![iclouddrive](https://user-images.githubusercontent.com/7400768/233445975-f16b1527-8a0b-467a-98f0-8594f430c6f8.png)

Abaixo outro exemplo, mas de uma nuvem da Google:
![diveMobileBackups](https://user-images.githubusercontent.com/7400768/233445996-7cbae8a7-0faa-4c7b-a257-a9c77d6a29cb.png)

Desde já agradeço pela atenção!

**Comment by lfcnassif** [comment]
The WhatsApp database (ChatStorage.sqlite) in the iCloud backup seems encrypted (enc suffix). You would need the encryption key to decrypt it. Anyway, even with the encryption key, currently IPED doesn't support decrypting that database, it needs the database already decrypted to extract the chats, calls, etc.

**Reply by dafeol** [reply]
Há alguma solução que substitua o Cellebrite para descriptografar esse arquivo antes de fazer a indexação?
Outra dúvida... tendo em vista q o Celebritte consegue realizar esse processo,  a chave para descriptografar vem junto aos arquivos baixados da nuvem?

**Reply by lfcnassif** [reply]
I have never had access to a cloud data warrant return, so unfortunately I can't give you a precise answer. My **guess** is that the WhatsApp encryption key is not in the cloud (at least in plain form), otherwise that would be a security flaw. I **guess** you would need access to the physical mobile phone, from where you may be able to extract the encryption key with a FFS extraction. Or maybe the Cellebrite software is able to do a password attack on the encryption key, not sure... Maybe other users can help you better than me on this subject, sorry.

**Reply by dafeol** [reply]
Entendi! Muito obrigado pela atenção

**Reply by Jefresongil** [reply]
EM alguns processos em que tive a oportunidade de trabalhar foram recebidos dados contidos em Nuvem de Google e/ou Apple, e haviam conteúdos trazidos em texto claro. Li em uma postagem da Whatsapp LLC, que em outubro/22 teria sido implementado uma senha em sua configuração de forma que o usuário poderia subir seu backup com senha. Então, pelas consultas que realizei, a chave Key seguia junto com o backup realizado na nuvem. Posso estar enganado. Mas entendo que era assim.

**Comment by rafael844** [comment]
Hi Nassif.
This .enc in icloudrive use to be encrypted, as you said, and even cellebrite cant decrypt. But sometimes ChatStorage.sqlite is in the backup folder, and Cellebrite only needs the FileInfoList.txt that comes with the package.

![image](https://github.com/sepinf-inc/IPED/assets/31928478/df6dc62d-242c-4a23-b844-f07025a13450)



Inside the FileInfoList.txt it has, as example:


```
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>relativePath</key>
	<string>ChatStorage.sqlite</string>
	<key>birth</key>
	<string>6834530376009</string>
	<key>groupID</key>
	<string>503451</string>
	<key>mode</key>
	<string>33134588</string>
	<key>modified</key>
	<string>6833453687758</string>
	<key>statusChanged</key>
	<string>683654587758</string>
	<key>userID</key>
	<string>5401</string>
	<key>size</key>
	<string>3915666451</string>
	<key>sizeBeforeCopy</key>
	<string>10956666800</string>
</dict>
</plist>","contents":{"signature":"BCWDLkmXBgE2tM2f34356P4wLD","owner":"20728237465487862341436","size":3915455776,"reference_signature":"AXxfQ168SNeu8H9Yfm0","wrapping_key":"/lNnTOZhCul6kdvcyQdnhKqqIUWX4=","pcsInfo":"7BdgfhfdsHDSKJHLSH3P2Z3+SSl7gutBa5BKMrfJAHSJbJJmbYTVNZicMpXWWwA=="},"___createdBy":{"id":"_2bc76f56cd27498e3456596e8993azazec","type":"WEAK"},"___modifiedBy":{"id":"_2bc76f56cd274678698e96e8993azazec","type":"WEAK"},"_db_generationNum":0,"fileType":0,"___modTime":9780000},"created":{"timestamp":1661995411171,"user":"_2bc76f5698e96e32377a768993ezazac","device":"1"},"modified":{"timestamp":978000,"user":"_2b76f56cd27498e96e32377a768993ec","device":"1"},"plugin_fields":{},"conflictLosersEtags":[]}][{"id":"F:08534FA4-97E0-45B1-AFE0-E6AA57018777:N0Hbft3OLzKTFuguaWkzZpJKqxQ=:7436:bhPK","recordType":"File","tombstone":false,"etag":"0","fields":{"___etag":"0","extension":"sqlite","protectionClass":3,"___recordID":{"id":"F:08345534FA4-934E0-0-E3453465477:N0Hbftfdg3OWkzZpJKqxQ=:7436:bhPK","type":"WEAK"},"version":1,"___createTime":1661995411170,"encryptedAttributes":"<?xml version="1.0" encoding="UTF-8"?>
```
IPED can process all files, .opus,. jpg, etc... inside the backup folder and those files are in this FileInfoList.txt file, but ChatStorage.sqlite it doesnt process. So we have to process first in PA and then in IPED.

**Reply by lfcnassif** [reply]
If ChatStorage.sqlite is unencrypted and not being decoded properly, regardless of its name (we detect it based on signature and table names, not on file name) please provide us with a sample not being decoded, so we can investigate.

**Reply by wladimirleite** [reply]
Or is `ChatStorage.sqlite` encrypted and Physical Analyzer uses the key in `FileInfoList.txt` to decrypt it?

**Comment by rafael844** [comment]
Thanks, I'll try to get a sample for testing.

**Reply by lfcnassif** [reply]
Have you succeed to get a sample iCloud warrant return, or just the ChatStorage.sqlite, that reproduces the issue/lack of support to share with us for testing?

> Or is ChatStorage.sqlite encrypted and Physical Analyzer uses the key in FileInfoList.txt to decrypt it?

Anyone can confirm if the encryption key is really into that file and where?

Or maybe the cause of IPED not decoding unencrypted ChatStorage.sqlite is related to this ticket #1294 I forgot.

If ChatStorage.sqlite is compressed using LZFSE, IPED is not able to decode it today. Again, I would need samples to investigate and test a future implementation.

**Reply by rafael844** [reply]
> Have you succeed to get a sample iCloud warrant return, or just the ChatStorage.sqlite, that reproduces the issue/lack of support to share with us for testing?

Not yet, we asked for it and our superiors are trying to get the aproval to share.

> Anyone can confirm if the encryption key is really into that file and where?

The only thing I can say is that without that file Cellebrite cant show  ChatStorage.sqlite

**Reply by lfcnassif** [reply]
Cellebrite software needs the original filenames (or paths) to work. That's why it needs that file, at least.

**Comment by tlragazzan** [comment]
A título de informação, se não for fornecido o arquivo FileInfoList.txt o Cellebrite não consegue decodificar o ChatStorage

**Reply by lfcnassif** [reply]
IPED doesn't need that file. If it is not working, we need a ChatStorage.sqlite sample not being decoded to investigate.

**Reply by tlragazzan** [reply]
FileInfoList.txt contains various information about Apple backup files, including file names.
All files are provided with encoded names, and this relationship is contained in FileInfoList.txt.

**Example**
File name: F_24118AE9-51F2-4B69-A478-CD54D181F5EC_1VsNrVJaX7qCl2EmPDJ8f2vF6pk=_6075083_PJNp
Name FileInfoList.txt : ChatStorage.sqlite

Specifically ChatStorage.sqlite, and other Whatsapp files and other databases are encoded in LFZSE. Sometimes, in addition to LZFSE, they are also encoded in LZIP.

I'm working on a solution, adopting the strategy of collecting all filenames from FileInfoList.txt, decoding the ones that need it and exporting it to a ZIP.

**Reply by lfcnassif** [reply]
As I said before, IPED WhatsApp decoding doesn't need original filename or path to work. If LZFSE or LZIP encoding are used, I need <b>samples</b> to implement and test a future support for them. The few WhatsApp samples from the cloud I have are not LZFSE encoded.

**Comment by lfcnassif** [comment, marked as answer]
Got 01 sample iCloud data with a few LZFSE files from André (PC/MG), thank you! Decompressed them using RagingMoose library, referenced on #1294, then processed the resulting data with IPED. WhatsApp conversations were decoded fine! Attachments were also linked properly, with no file renaming/mapping from `FileInfoList.txt`, that's not needed by our WhatsApp parser, as I said before. Interestingly, most files in the sample set are not encoded, just a few of them are LZFSE compressed, including ChatStorage.sqlite and other sqlite databases, as @tlragazzan said.

I should have a working draft in the next days. But for a final implementation, I would need more samples to properly test the feature, for example, with LZIP compression. @tlragazzan, do you mean some files are LZFSE compressed and others LZIP compressed? Or some files really are LZIP compressed after LZFSE compression, or the opposite? This seems strange to me... If you or anyone else could provide more samples, that would be very useful to finish the feature.

**Reply by lfcnassif** [reply]
For those that can help testing, here is a snapshot with a working draft support:
https://github.com/sepinf-inc/IPED/actions/runs/5843809106

There is still a limitation, to be addressed in the next days, described here:
https://github.com/sepinf-inc/IPED/pull/1812

But I think it should work for most purposes.

**Reply by lfcnassif** [reply]
I still need LZIP compressed samples to try to handle them properly, I don't have any.

**Reply by tlragazzan** [reply]
Some files are LZIP compressed after LZFSE.

**Reply by lfcnassif** [reply]
Could you share some samples privatelly with me? My email is public in my profile.

**Comment by rafael844** [comment]
> For those that can help testing, there is already a snapshot with a working draft here:
https://github.com/sepinf-inc/IPED/actions/runs/5843809106

We tested it and can confirm that the snapshot is working and bringing the whatsapp conversations, including linking media files.

**Reply by lfcnassif** [reply]
Thank you @rafael844 for testing! @dafeol since you were interested first and opened this discussion, could you also test the snapshot please?

**Reply by tlragazzan** [reply]
I tested it with several icloud backups and it is bringing the whatsapp conversations, linking the media files to the conversations.

I'm running more tests.

Please disregard the LZIP compressed files, I reviewed the original files and they are in LZFSE, during some processing this LZIP file came up.

Sorry =(

**Reply by lfcnassif** [reply]
Thanks @tlragazzan for all tests and for warning about LZIP compression absence, so I'll close #1813.

**Reply by dafeol** [reply]
Acabo de testar, com um backup com o qual já havia feito a indexação através do IPED 4.1.2. O snapshot disponibilizado para teste conseguiu nesta nova indexação realizar a leitura do backup do WhatsApp, trazendo todos os diálogos montados, inclusive de conversas que aparentemente são de grupos. 
Parabéns a todos os envolvidos! Esse recurso vai trazer um imensurável ganho nas próximas análises!

![teste](https://github.com/sepinf-inc/IPED/assets/7400768/75ae6f5c-db4f-49d1-b3a4-51b5ba272cc7)
![teste2](https://github.com/sepinf-inc/IPED/assets/7400768/dc69a864-092e-4266-b3a2-8e8852c536b5)

**Reply by dafeol** [reply]
Ainda analisando a indexação gerada pelo snapshot disponibilizado, notei que alguns arquivos, que conseguia identificar como sendo trocados entre a conta do backup com um segundo interlocutor, ele não conseguiu puxar e vincular. Quando utilizo como parâmetro de pesquisa a conta com quem ele trocou arquivos (Ex.: **5581XXXXXXXX**@s.whatsapp.net) ele não encontra todos os arquivos trocados, encontrando somente quando pesquiso diretamente pelo nome do arquivo (Que identifiquei na indexação com a versão anterior, a oficial)

![teste3](https://github.com/sepinf-inc/IPED/assets/7400768/14382dcf-8be4-40f4-8875-f56aedb6ff8b)

**Reply by lfcnassif** [reply]
Thanks for testing @dafeol. So you used the exact same search criteria alone, and it brought less results? Could you check one of those mp4 or opus files returned before and compare the content displayed into Text viewer tab if it has any difference? The snapshot changes shouldn't affect mp4 or opus processing. Those files are present in the case processed by the snapshot, right?

**Reply by dafeol** [reply]
Correto. Utilizando o mesmo parâmetro de pesquisa, no caso terminal que identifica o usuário do WhatsApp, com o snapshot ele me retorna uma quantidade menor de arquivos. Não encontrando por exemplos os arquivos de mídia (.opus, .pdf e .mp4):

![image](https://github.com/sepinf-inc/IPED/assets/7400768/5e4ce9cd-8633-4b5e-a3f2-b962f3de5d85)

Se eu buscar o arquivo que encontrei vinculado ao diálogo entre dois interlocutores na primeira extração (IPED v.4.1.2), na extração realizada pelo snapshot disponibilizado, pelo nome dele, eu até encontro o arquivo, mas reparei que ele não vem com as informações da coluna "EMBEDDEDRELATIONSHIPID", mas os restantes das informações do arquivo, aparentemente, são as mesmas

![image](https://github.com/sepinf-inc/IPED/assets/7400768/31613c49-cb0f-432c-a786-6d21f05de182)

Quanto as informações exibidas na guia do visualizador de texto... não sei se olhei na área certa, mas lá tbm notei que falta a informação de embeddedRelationshipId:

![image](https://github.com/sepinf-inc/IPED/assets/7400768/d5426db2-8855-4ccb-a271-3980eb2dcc48)

**Reply by wladimirleite** [reply]
I am not following this discussion close enough, but I think SHA-256 needs to be enabled for WhatsApp chat attachments to be linked properly.
@dafeol, can you check in your case (if items have SHA-256 in the metadata tab) or if it was enabled (in the processing log)?

**Reply by lfcnassif** [reply]
What's the full path of the above selected file in the case processed using the snapshot version? In the 4.1.2 case, it was into a TAR archive, having just the modified date. In the 4.2-snapshot case, I can't see the full path in the picture, but it has all 3 file system dates, so I guess it is not into an archive, like TAR, this would explain the difference, the embeddedRelationShipId property is present in files extracted from containers. It is important to process the exact same input and with the same processing options when comparing different versions.

**Reply by lfcnassif** [reply]
If you processed with `ignoreDuplicates = true` that can also lead to this kind of differences: first case took the file from a container and the second case could have picked up a duplicate from outside the container.

**Reply by dafeol** [reply]
Assim que der aqui, tentarei refazer as indexações, utilizando os mesmo parâmetros de configuração, incluindo a ativação do SHA-256, como o colega @tc-wleite observou também.

**Reply by dafeol** [reply]
Mas uma dúvida... para a indexação dessas nuvens então o ideal é não mudar a configuração "ignoreDuplicates" ? Deixar ela com o valor default (false) ?

**Reply by lfcnassif** [reply]
I don't recommend enabling `ignoreDuplicates` for most cases as explained above that option in IPEDConfig.txt. Actually I thought to remove it because it is inherently unsafe, see:
https://github.com/sepinf-inc/IPED/discussions/931

But ended keeping it because it might be useful when you have to process dozens of backups from the same user/machine, which is not the case here.

**Comment by lfcnassif** [comment]
Hello all,

Just finished my planned enhancement to better detect WhatsApp iOS account info, to be independent of a specific plist file name. Now the owner account should be detected fine and will be shown in all chat balloons to the right in WA conversations, and also displayed fine in the graph analysis. Snapshot will be ready here in 10 minutes:
**edited**: https://github.com/sepinf-inc/IPED/actions/runs/5871612960

Please give it a try. After some of you test it and report results are fine, I'll integrate the new feature in main version.

**Reply by dafeol** [reply]
Testei aqui, observando as configurações que sugeriram (SHA-256 e ignoreDuplicates = fase), não sei se continuo errando algo na hora de indexar, mas o erro continua o mesmo reportado na mensagem anterior (sem as informações da coluna/campo "EMBEDDEDRELATIONSHIPID" e quando procuro utilizando o terminal de uma conta do WhatsApp como parâmetro, com quem a conta principal se comunicou, ele não me traz todos os arquivos trocados, só consigo localizá-los quando procuro diretamente pelo nome do arquivo que quero e mesmo encontrando estes arquivos, não os vejo vinculados ao diálogo do WhatsApp montado)

**Reply by lfcnassif** [reply]
Could you please send a screenshot of your SHA-256 configuration? The right one is below. About searching for a WhatsApp number, it is not the direct way to find shared files, you should click on the returned Chat with that number, then all shared files should be listed into the References tab, but that will work only if the sha-256 configuration in correct into conf/HashTaskConfig.txt, should be lowercase:
![image](https://github.com/sepinf-inc/IPED/assets/7276994/85cc6dae-88f1-4813-a470-67ccc260176a)

About EMBEDDEDRELATIONSHIPID, it was indexed in your 4.1.2 case because the input data was compressed. I asked for the full path of mp4/opus files in the new case to help you diagnose, and I'm still waiting for the info...

**Reply by lfcnassif** [reply]
So, see file paths are different, in first case the file is inside a TAR archive, in second not. That explains the absence of EMBEDDEDRELATIONSHIPID property in the second case, as I said before. If you search for path:XXXXXX where XXXXX is your number, those files should be returned in the second case.

And please click on "Propriedades Avançadas" in the Metadata viewer and check if sha-256 was computed. Maybe you used another processing profile with a wrong configuration.

**Reply by rafael844** [reply]
We tested here and its working as expected. Contact name is showing correctly. Thank you @lfcnassif .

**Reply by lfcnassif** [reply]
Thanks for testing!

**Reply by lfcnassif** [reply]
I merged the feature yesterday and I plan to release the new version today.

**Comment by dafeol** [comment]
![image](https://github.com/sepinf-inc/IPED/assets/7400768/3c86b475-68a1-45b3-8c36-d580aaa41375)

Obs: Apaguei a postagem anterior pq faltou ocultar alguns dados

**Reply by lfcnassif** [reply]
Hum... sha-256 was computed properly. If attachments are not being linked to chats in your case, sorry, I'm out of ideas, I would need you to send your data privately to me, so I can try to reproduce and investigate, it doesn't happen to me, neither to @rafael844, neither to @tlragazzan.

**Reply by dafeol** [reply]
Eu verifiquei e realmente as indexações tinham aquelas diferenças observadas (indexei de origens diferentes, onde uma tinha o arquivos descompactados e a outra não). Também consegui localizar a aba específica, que você apontou como a ideal para identificar os arquivos trocados entre determinados interlocutores do chat (Aba Referências). Porém neste último ponto ainda ficou uma dúvida, pq além dos arquivos que eu encontro nesta aba "Referencias", ainda encontro arquivos que, ao menos da minha leitura, aparenta ter sido trocada entre os mesmo interlocutores (Não sei se por não constar diretamente na pasta de backup, mas sim na pasta iclouddrive por exemplo). 
Daí meu questionamento é: Além dos arquivos encontrados nesta guia, os que eu encontrar diretamente em por exemplo "Multimídia > Imagens", eu poderei afirmar que foram trocados entre os dois interlocutores, mesmo não presentes lá no chat?  A seguinte interpretação está correta:

![image](https://github.com/sepinf-inc/IPED/assets/7400768/09d54e3e-1e1b-4b84-b25c-219a1bbdf54f)

Quanto ao pedido dos dados para sua análise, como seria feito? Pergunto pq os arquivos que tenho em mãos ainda estão em fase de análise. Não tenho nenhum outro de investigação já concluída disponível no momento.

**Reply by lfcnassif** [reply]
Files shown in References tab are files the tool detected were sent or received for sure, they are into the chat body. Maybe the files you found are related to deleted messages, not available in WhatsApp database anymore. Maybe user created those files and put into the WhatsApp folder using a similar name pattern. Anyway, the tool just links attachments still referenced by the WhatsApp database.
